### PR TITLE
ENH: stats.truncate: truncated random variables

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-9439

#### What does this implement/fix?
Adds the ability to truncate the distribution underlying a random variable. Truncation can be applied before or after other transformations, which makes it easier to get the desired result.

```python3
import matplotlib.pyplot as plt
from scipy.stats import Normal, truncate

X = Normal()

# Truncate shifted/scaled variable (new)
Y = truncate(2*X + 1, lb=-2, ub=2)
Y.plot()
plt.show()
```
image

```python3
# Shift/scale truncated variable (like `truncnorm`)
Y = 2*truncate(X, lb=-2, ub=2) + 1
Y.plot()
plt.show()
```
image

#### Additional information
This will not be quite as robust as `truncnorm` as written, but I can look at improving it / testing it with `truncnorm` tests as a followup.

### CODE REVIEW

**Review by steppi — APPROVED** [review]
It would be nice to improve this to choose the tail worked with to get better numerical results, but for now, I think it's fine just to get the interface settled.

Also, I noticed that the default of sorting parameters in the `__repr__` leads to slightly odd results here. We have

```python
import scipy.stats as stats

X = Normal(mu=0, sigma=1)
Y = truncate(X, -1, 1)
print(Y)
# TruncatedNormal(lb=-1.0, mu=0.0, sigma=1.0, ub=1.0)
```

whereas ideally I think the conceptually related `lb` and `ub` should be placed next to each other. Especially since they are abbreviated. A quick fix would be to just override the `__repr__` for `TruncatedDistribution`.

I agree with the choice to have a function `truncate`. It seems consistent with the interface for other transformations, where we overload operators or built-ins when possible, and use functions otherwise, allowing for easily composable operations on distributions / random variables.

**Review by tupui — APPROVED** [review]

File: scipy/stats/_distribution_infrastructure.py [hunk 1]
```diff
@@ -3676,6 +3677,146 @@ def __repr__(self):
...
+
+    _ub_domain = _RealDomain(endpoints=('lb', inf), inclusive=(False, True))
+    _ub_param = _RealParameter('ub', symbol=r'b_u',
+                                  domain=_ub_domain, typical=(0.8, 0.9))
+
+    _parameterizations = [_Parameterization(_lb_param, _ub_param),
+                          _Parameterization(_lb_param),
+                          _Parameterization(_ub_param)]
+
+    def __init__(self, dist, *args, lb=-np.inf, ub=np.inf, **kwargs):
```
Comment (jorenham) [inline scipy/stats/_distribution_infrastructure.py:3699]:
```suggestion
    def __init__(self, dist, /, *args, lb=-np.inf, ub=np.inf, **kwargs):
```

Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:3699]: Sure, will do. And `dist` will be `X`.

**Comment by mdhaber** [comment]
> It would be nice to improve this to choose the tail worked with to get better numerical results, but for now, I think it's fine just to get the interface settled. 

Yup, I mentioned that in the code comments.

> Also, I noticed that the default of sorting parameters in the __repr__ leads to slightly odd results here.

Yeah, I planned to play around with `__repr__` some more when we can play with these composable transformations together. Another big improvement would be for the `__repr__` of a class created with `make_distribution` to name the underlying distribution rather than just being `CustomDistribution`.

File: scipy/stats/_distribution_infrastructure.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:3739]:
```suggestion
    def __init__(self, X, /, *args, lb=-np.inf, ub=np.inf, **kwargs):
```

File: scipy/stats/_distribution_infrastructure.py [hunk 2]
```diff
@@ -3717,6 +3717,146 @@ def __repr__(self):
...
+        return super().__init__(dist, *args, lb=lb, ub=ub, **kwargs)
```
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:3740]:
```suggestion
        return super().__init__(X, *args, lb=lb, ub=ub, **kwargs)
```

**Comment by tylerjereddy** [comment]
The description here was sufficiently concise that I took a stab at writing the release note myself, but you may to check it (only available locally for now...).

**Pull request merged**